test/e2e/storage/volume_metrics.go

```diff
@@ -165,6 +166,123 @@ var _ = utils.SIGDescribe("[Serial] Volume metrics", func() {
+					HostPath: &v1.HostPathVolumeSource{Path: "/data"},
+				},
+				NamePrefix:       "pv-test-",
+				StorageClassName: className,
+			}
+			pvcConfig = framework.PersistentVolumeClaimConfig{StorageClassName: &className}
+		)
+
+		const (
+			classKey     = "storage_class"
+			namespaceKey = "namespace"
+
+			boundPVKey    = "pv_collector_bound_pv_count"
+			unboundPVKey  = "pv_collector_unbound_pv_count"
+			boundPVCKey   = "pv_collector_bound_pvc_count"
+			unboundPVCKey = "pv_collector_unbound_pvc_count"
+		)
+
+		// validator used to validate each metric's values, the length of metricValues
+		// should be 4, and the elements should be bound pv count, unbound pv count, bound
+		// pvc count, unbound pvc count in turn.
+		validator := func(metricValues []map[string]int64) {
+			Expect(len(metricValues)).To(Equal(4),
+				"Wrong metric size: %d", len(metricValues))
+
+			controllerMetrics, err := metricsGrabber.GrabFromControllerManager()
+			Expect(err).NotTo(HaveOccurred(), "Error getting c-m metricValues: %v", err)
+
+			metrics := []struct {
+				name      string
+				dimension string
+			}{
+				{boundPVKey, classKey},
+				{unboundPVKey, classKey},
+				{boundPVCKey, namespaceKey},
+				{unboundPVCKey, namespaceKey},
+			}
+			for i, metric := range metrics {
+				expectValues := metricValues[i]
+				if expectValues == nil {
+					expectValues = make(map[string]int64)
+				}
+				actualValues := getPVControllerMetrics(controllerMetrics, metric.name, metric.dimension)
+				Expect(actualValues).To(Equal(expectValues),
+					"Wrong pv controller metric %s(%s): wanted %v, got %v",
+					metric.name, metric.dimension, expectValues, actualValues)
+			}
+		}
+
+		BeforeEach(func() {
+			pv = framework.MakePersistentVolume(pvConfig)
+			pvc = framework.MakePersistentVolumeClaim(pvcConfig, ns)
+		})
+
+		AfterEach(func() {
+			if err := framework.DeletePersistentVolume(c, pv.Name); err != nil {
+				framework.Failf("Error deleting pv: %v", err)
+			}
+			if err := framework.DeletePersistentVolumeClaim(c, pvc.Name, pvc.Namespace); err != nil {
+				framework.Failf("Error deleting pvc: %v", err)
+			}
+		})
+
+		It("should create none metrics for pvc controller before creating any PV or PVC", func() {
+			if !metricsGrabber.HasRegisteredMaster() {
+				framework.Skipf("Environment does not support getting controller-manager metrics - skipping")
+			}
+			validator([]map[string]int64{nil, nil, nil, nil})
+		})
+
+		It("should create unbound pv count metrics for pvc controller after creating pv only",
+			func() {
+				var err error
+				if !metricsGrabber.HasRegisteredMaster() {
+					framework.Skipf("Environment does not support getting controller-manager metrics - skipping")
+				}
+				pv, err = framework.CreatePV(c, pv)
+				Expect(err).NotTo(HaveOccurred(), "Error creating pv: %v", err)
+				waitForPVControllerSync(metricsGrabber, unboundPVKey, classKey)
+				validator([]map[string]int64{nil, {className: 1}, nil, nil})
```

Comment:
These metrics appear to be checking absolute number of bound or unbound PVs. Will this not fail when some other PV might exist in the cluster while this test is running?

Reply:
All volume metric e2e tests are labeled as `[Serial]`(see [here](https://github.com/kubernetes/kubernetes/blob/35ed5338b104f41c91efb2b842e92fded6d0b509/test/e2e/storage/volume_metrics.go#L38)), so we can consider that no other existing PVs while this test is running.

Reply:
Yes - I know but reality is stranger than that. I have fixed a number of flakes in this test suite because something else caused metrics to jump around. We have to be careful, and only observe increment in metric values, rather than absolute values - because asserting on absolute values is almost sure to be error prone.

Reply:
ACK. This is indeed a problem as some other tests maybe create PV/PVCs and forget to cleanup. I will change to use increment instead of absolute values.

Reply:
@gnufied The e2e tests are updated to validate the relative increment value instead of absolute value, PTAL, thanks.